Add support for default fallback values

a-alamoudi opened the pull request:
Addresses #11, to add support for fallback default values for variables

Providing values for all variables can be cumbersome at times when there are many of them. This feature is meant to allow users to pass in a default value to fall back on in case the user doesn't provide one, all specified in the same config file. 

To preserve backwards compatibility, a new method `EnableSubstitutionsWithDelimitedFallbackDefaults`, allows specifying a delimiter that separates the variable from the fallback default value.

a-alamoudi committed:
Add support for default fallback values

a-alamoudi committed:
Add support for urls as default values

- [UPDATE] Only capture text after the first delimiter occurrence (e.g., in urls with '').

molinch commented on ConfigurationSubstitutor/ConfigurationSubstitutor.cs:
```diff
@@ -68,7 +68,7 @@ private string ApplySubstitution(IConfiguration configuration, string value, Has
                 if (substitutedValue == null && !string.IsNullOrEmpty(_fallbackDefaultValueDelimiter))
                 {
-                    var delimitedVals = capture.Value.Split(new[] { _fallbackDefaultValueDelimiter }, StringSplitOptions.None);
+                    var delimitedVals = capture.Value.Split(new[] { _fallbackDefaultValueDelimiter }, 2, StringSplitOptions.None);
```
What's the reason for using 2 here?

a-alamoudi replied:
That's just to make sure that only the first occurrence of the delimiter is used to split the captured value string into exactly two parts. So, in the case where multiple occurrences of the specified delimiter are present, only the first occurrence would be used to tokenize the string. A common example of that is passing URLs as a default value along with `:` as a delimiter; for instance:

```
$(VAR1:http://example.com)
```

There the split should take place at the first occurrence of `:` of the captured value string (i.e., `VAR1:http://example.com`), and return to us the second part fully intact as `http://example.com` and not `http`.

molinch commented on ConfigurationSubstitution.Tests/ConfigurationTests.cs:
```diff
@@ -370,5 +370,187 @@ public void Should_get_substituted_value_when_using_long_substituable_pattern(Fu
+                {
+                    { "Foo", "$(Var1:Testo)" },
+                })
+                .EnableSubstitutionsWithDelimitedFallbackDefaults("$(", ")", null);
```
Question: wouldn't it make more sense that in both `.EnableSubstitutions` and `.EnableSubstitutionsWithDelimitedFallbackDefaults` we always throw if any argument is null or empty?

a-alamoudi replied:
Good point. Yeah, I think that makes perfect sense. This highlighted test only looks at the `fallbackDefaultValueDelimiter` argument, but I think think the remaining two arguments, namely `substitutableStartsWith` and `substitutableEndsWith`, should be included too and throw an exception when they're empty or null.

molinch replied:
If you could do that directly in the extension method then, failing fast is good
PR quality is really good, thanks a lot!

a-alamoudi replied:
Cool! Will do and commit that asap
Thank you very much for taking the time to review and build this awesome project, I'm glad I could contribute to it 🙂

a-alamoudi replied:
Just committed  e3b7ed9e0ca8050ff7498fb2203aa0062b7c4b58, which has the discussed suggestions along some unit tests to verify the correct behaviour. Please take a look and see if anything is missing.

a-alamoudi committed:
Throw exceptions with faulty constructor arguments

- [UPDATE] Throw ArgumentException/ArgumentNullException when null or empty strings are passed to the ConfigurationSubstitutor constructor.
- [TEST] Add new unit tests for both EnableSubstitutions and EnableSubstitutionsWithDelimitedFallbackDefaults when faulty arguments are passed.

molinch commented:
Thanks @a-alamoudi really great stuff 👍 🥇 💯 
I'll do a nuget push soon
